### PR TITLE
[Mxplayer] Fix extractor

### LINKED ISSUE #4946: MX player no longer working
### DO NOT REMOVE OR SKIP THE ISSUE TEMPLATE

- [X] I understand that I will be **blocked** if I remove or skip any mandatory\* field

### Checklist

- [X] I'm reporting a broken site
- [X] I've verified that I'm running yt-dlp version **2022.09.01** ([update instructions](https://github.com/yt-dlp/yt-dlp#update)) or later (specify commit)
- [X] I've checked that all provided URLs are playable in a browser with the same IP and same login details
- [X] I've checked that all URLs and arguments with special characters are [properly quoted or escaped](https://github.com/yt-dlp/yt-dlp/wiki/FAQ#video-url-contains-an-ampersand--and-im-getting-some-strange-output-1-2839-or-v-is-not-recognized-as-an-internal-or-external-command)
- [X] I've searched the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues **including closed ones**. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)
- [X] I've read about [sharing account credentials](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#are-you-willing-to-share-account-details-if-needed) and I'm willing to share it if required

### Region

_No response_

### Provide a description that is worded well enough to be understood

https://www.mxplayer.in/

This site is not working anymore. Needs update.
Previously it worked fine. Now it seems broken.

I use yt-dlp for streaming videos in external players like MPC-HC.
Screenshot (648)

### Provide verbose output that clearly demonstrates the problem

- [X] Run **your** yt-dlp command with **-vU** flag added (`yt-dlp -vU <your command line>`)
- [X] Copy the WHOLE output (starting with `[debug] Command-line config`) and insert it below

### Complete Verbose Output

```shell
[debug] Command-line config: ['-vU', 'https://www.mxplayer.in/movie/watch-kitne-door-kitne-paas-movie-online-a9e9c76c566205955f70d8b2cb88a6a2?watch=true']
[debug] Encodings: locale cp1252, fs utf-8, pref cp1252, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version 2022.09.01 [5d7c7d6] (win32_exe)
[debug] Python 3.8.10 (CPython 64bit) - Windows-10-10.0.19044-SP0
[debug] Checking exe version: ffmpeg -bsfs
[debug] Checking exe version: ffprobe -bsfs
[debug] exe versions: none
[debug] Optional libraries: Cryptodome-3.15.0, brotli-1.0.9, certifi-2022.06.15, mutagen-1.45.1, sqlite3-2.6.0, websockets-10.3
[debug] Proxy map: {}
[debug] Loaded 1670 extractors
[debug] Fetching release info: https://api.github.com/repos/yt-dlp/yt-dlp/releases/latest
Latest version: 2022.09.01, Current version: 2022.09.01
yt-dlp is up to date (2022.09.01)
[debug] [Mxplayer] Extracting URL: https://www.mxplayer.in/movie/watch-kitne-door-kitne-paas-movie-online-a9e9c76c566205955f70d8b2cb88a6a2?watch=true
[Mxplayer] watch-kitne-door-kitne-paas-movie-online: Downloading JSON metadata
ERROR: [Mxplayer] a9e9c76c566205955f70d8b2cb88a6a2: Unable to download JSON metadata: HTTP Error 400: Bad Request (caused by <HTTPError 400: 'Bad Request'>); please report this issue on  https://github.com/yt-dlp/yt-dlp/issues?q= , filling out the appropriate issue template. Confirm you are on the latest version using  yt-dlp -U
  File "yt_dlp\extractor\common.py", line 670, in extract
  File "yt_dlp\extractor\mxplayer.py", line 135, in _real_extract
  File "yt_dlp\extractor\common.py", line 1030, in download_content
  File "yt_dlp\extractor\common.py", line 994, in download_handle
  File "yt_dlp\extractor\common.py", line 864, in _download_webpage_handle
  File "yt_dlp\extractor\common.py", line 821, in _request_webpage

  File "yt_dlp\extractor\common.py", line 803, in _request_webpage
  File "yt_dlp\YoutubeDL.py", line 3656, in urlopen
  File "urllib\request.py", line 531, in open
  File "urllib\request.py", line 640, in http_response
  File "urllib\request.py", line 569, in error
  File "urllib\request.py", line 502, in _call_chain
  File "urllib\request.py", line 649, in http_error_default
urllib.error.HTTPError: HTTP Error 400: Bad Request
```

### PR DESCRIPTION
**Pull request by itachi-19**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

</details>

I figured out the latest API url and parameters of the website's streams and fixed the extractor accordingly. Since the tests were failing due to outdated API, I updated the tests. Please note this only fixes MxplayerIE and <b>NOT</b> MxplayerShowIE, which I'll cover in the future, since I don't know the correct API url for it currently.

Fixes #4946

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

### CODE REVIEW

File: yt_dlp/extractor/mxplayer.py [hunk 1]
```diff
@@ -1,178 +1,99 @@
 from .common import InfoExtractor
 from ..compat import compat_str
 from ..utils import try_get
+import re
 
 
 class MxplayerIE(InfoExtractor):
     _VALID_URL = r'https?://(?:www\.)?mxplayer\.in/(?P<type>movie|show/[-\w]+/[-\w]+)/(?P<display_id>[-\w]+)-(?P<id>\w+)'
     _TESTS = [{
-        'url': 'https://www.mxplayer.in/show/watch-my-girlfriend-is-an-alien-hindi-dubbed/season-1/episode-1-online-9d2013d31d5835bb8400e3b3c5e7bb72',
```
Comment (elyse0) [inline yt_dlp/extractor/mxplayer.py:9]: Don't delete tests, if these urls are not longer available you can add

```python
{
 ....,
 'skip': 'No longer available',
}
```

File: yt_dlp/extractor/mxplayer.py [hunk 2]
```diff
@@ -1,178 +1,99 @@
...
+
+        if video_type == 'episode':
+            series = data_json.get('container').get('container').get('title')
+            season = data_json.get('container').get('title')
+            episode_number = data_json.get('sequence')
+
+            if season:
+                match = re.match(r'Season (?P<season_number>\d+)', season)
+                if match:
+                    season_number = int(match.group('season_number'))
```
Comment (elyse0) [inline yt_dlp/extractor/mxplayer.py:70]: I think you could use `self._search_regex()`

File: yt_dlp/extractor/mxplayer.py (hunk 1 above)
Comment (pukkandan) [reply yt_dlp/extractor/mxplayer.py:9]: The old test still works in browser. No clue why OP removed it

File: yt_dlp/extractor/mxplayer.py (hunk 2 above)
Comment (itachi-19) [reply yt_dlp/extractor/mxplayer.py]: Done.

File: yt_dlp/extractor/mxplayer.py [hunk 3]
```diff
@@ -1,178 +1,99 @@
...
-                'width': thumb.get('width'),
-                'height': thumb.get('height'),
-            })
+        video_type, display_id, video_id = self._match_valid_url(url).groups()
+
+        if 'show' in video_type:
+            video_type = 'episode'
+
+        API_URL = f'https://api.mxplay.com/v1/web/detail/video?type={video_type}&id={video_id}'
+        STREAM_URL = 'https://llvod.mxplay.com/{}'
```
[the remark this answers is not in the document]
Comment (itachi-19) [reply yt_dlp/extractor/mxplayer.py]: Done.

File: yt_dlp/extractor/mxplayer.py (hunk 1 above)
Comment (itachi-19) [reply yt_dlp/extractor/mxplayer.py:9]: Re-added the removed tests. Some URLs are broken and some don't play even on Firefox or Edge with message - "Video cannot be played on this browser".

File: yt_dlp/extractor/mxplayer.py [hunk 4]
```diff
@@ -1,178 +1,99 @@
...
 
+        data_json = self._download_json(API_URL, display_id)
+
+        title = data_json['title']
+        duration = data_json.get('duration')
```
[the remark this answers is not in the document]
Comment (itachi-19) [reply yt_dlp/extractor/mxplayer.py]: Done.

File: yt_dlp/extractor/mxplayer.py [hunk 5]
```diff
@@ -1,178 +1,99 @@
...
-            elif dct.get('extension') == 'm3u8':
-                frmt, subs = self._extract_m3u8_formats_and_subtitles(dct.get('playUrl'), display_id, fatal=False)
-                formats.extend(frmt)
-                subtitles = self._merge_subtitles(subtitles, subs)
+        description = data_json.get('description')
+        series, season, season_number, episode_number = None, None, None, None
+
+        if video_type == 'episode':
+            series = data_json.get('container').get('container').get('title')
+            season = data_json.get('container').get('title')
```
[the remark this answers is not in the document]
Comment (itachi-19) [reply yt_dlp/extractor/mxplayer.py]: Done.

File: yt_dlp/extractor/mxplayer.py [hunk 6]
```diff
@@ -1,178 +1,99 @@
...
+            episode_number = data_json.get('sequence')
+
+            if season:
+                match = re.match(r'Season (?P<season_number>\d+)', season)
+                if match:
+                    season_number = int(match.group('season_number'))
+
+        for key in 'dash', 'hls':
```
[the remark this answers is not in the document]
Comment (itachi-19) [reply yt_dlp/extractor/mxplayer.py]: Done.

File: yt_dlp/extractor/mxplayer.py [hunk 7]
```diff
@@ -1,178 +1,99 @@
...
+            playlist_url = STREAM_URL.format(data_json['stream'][key]['high'])
+
+            if key == 'dash':
+                frmt, subs = self._extract_mpd_formats_and_subtitles(playlist_url, display_id, fatal=False)
+            else:
+                frmt, subs = self._extract_m3u8_formats_and_subtitles(playlist_url, display_id, fatal=False)
+
+            formats.extend(frmt)
+            subtitles = self._merge_subtitles(subtitles, subs)
```
[the remark this answers is not in the document]
Comment (itachi-19) [reply yt_dlp/extractor/mxplayer.py]: Done.

File: yt_dlp/extractor/mxplayer.py [hunk 8]
```diff
@@ -106,73 +108,99 @@ class MxplayerIE(InfoExtractor):
...
-                frmt, subs = self._extract_m3u8_formats_and_subtitles(dct.get('playUrl'), display_id, fatal=False)
-                formats.extend(frmt)
-                subtitles = self._merge_subtitles(subtitles, subs)
+        series, season, season_number, episode_number = None, None, None, None
+
+        if video_type == 'episode':
+            series = traverse_obj(data_json, ('container', 'container', 'title'))
+            season = traverse_obj(data_json, ('container', 'title'))
+            episode_number = data_json.get('sequence')
+            season_number = int(self._search_regex(r'Season (\d+)', season, 'Season Number'))
```
Comment (elyse0) [inline yt_dlp/extractor/mxplayer.py:177]: Make it non-fatal and use int_or_none()

```python
season_number = int_or_none(self._search_regex(r'Season (\d+)', season, 'Season Number', fatal=False))
```

Comment (elyse0) [inline yt_dlp/extractor/mxplayer.py:177]: By the way, can this be inline? If the values don't exist, we'll get None by default

File: yt_dlp/extractor/mxplayer.py [hunk 9]
```diff
@@ -106,73 +108,99 @@ class MxplayerIE(InfoExtractor):
...
-                'height': thumb.get('height'),
-            })
+        video_type, display_id, video_id = self._match_valid_url(url).groups()
+
+        if 'show' in video_type:
+            video_type = 'episode'
+
+        data_json = self._download_json(
+            f'https://api.mxplay.com/v1/web/detail/video?type={video_type}&id={video_id}', display_id
+        )
```
Comment (elyse0) [inline yt_dlp/extractor/mxplayer.py:167]: Avoid trailing parentheses

```python
 data_json = self._download_json(
            f'https://api.mxplay.com/v1/web/detail/video?type={video_type}&id={video_id}', display_id)
```

File: yt_dlp/extractor/mxplayer.py [hunk 10]
```diff
@@ -106,73 +108,99 @@ class MxplayerIE(InfoExtractor):
...
+        series, season, season_number, episode_number = None, None, None, None
+
+        if video_type == 'episode':
+            series = traverse_obj(data_json, ('container', 'container', 'title'))
+            season = traverse_obj(data_json, ('container', 'title'))
+            episode_number = data_json.get('sequence')
+            season_number = int(self._search_regex(r'Season (\d+)', season, 'Season Number'))
+
+        for stream_type in 'dash', 'hls':
+            playlist_url = 'https://llvod.mxplay.com/{}'.format(data_json['stream'][stream_type]['high'])
```
Comment (elyse0) [inline yt_dlp/extractor/mxplayer.py:180]: Use an f-string

```python
playlist_url = f'https://llvod.mxplay.com/{data_json["stream"][stream_type]["high"]}'
```

File: yt_dlp/extractor/mxplayer.py [hunk 11]
```diff
@@ -9,15 +10,15 @@ class MxplayerIE(InfoExtractor):
         'url': 'https://www.mxplayer.in/show/watch-my-girlfriend-is-an-alien-hindi-dubbed/season-1/episode-1-online-9d2013d31d5835bb8400e3b3c5e7bb72',
         'info_dict': {
             'id': '9d2013d31d5835bb8400e3b3c5e7bb72',
+            'display_id': 'episode-1-online',
             'ext': 'mp4',
             'title': 'Episode 1',
-            'description': 'md5:62ed43eb9fec5efde5cf3bd1040b7670',
+            'description': "Watch Season 1, Episode 1 of the show My Girlfriend Is an Alien (Hindi Dubbed).'",
```
Comment (elyse0) [inline yt_dlp/extractor/mxplayer.py:16]: Why did you change some descriptions from md5 to readable text? 🤔

File: yt_dlp/extractor/mxplayer.py [hunk 12]
```diff
@@ -28,53 +29,54 @@ class MxplayerIE(InfoExtractor):
...
             'id': 'b9fa28df3bfb8758874735bbd7d2655a',
+            'display_id': 'episode-1-online',
             'ext': 'mp4',
             'title': 'Knock Knock (Hindi Dubbed)',
-            'description': 'md5:b195ba93ff1987309cfa58e2839d2a5b',
-            'season_number': 0,
-            'episode_number': 0,
+            'description': 'Evan, a devoted husband, and father, invites double trouble as he allows two girls to seek shelter at his place. His sweet gesture takes the shape of a dangerous cat and mouse game.',
+            'season_number': None,
+            'episode_number': None,
```
Comment (elyse0) [inline yt_dlp/extractor/mxplayer.py:37]: If the value is None, you don't need to add that key to the test. I think... mmm

File: yt_dlp/extractor/mxplayer.py (hunk 11 above)
Comment (pukkandan) [reply yt_dlp/extractor/mxplayer.py:16]: This one is not too long. So it is fine

File: yt_dlp/extractor/mxplayer.py (hunk 12 above)
Comment (pukkandan) [reply yt_dlp/extractor/mxplayer.py:37]: Yes, don't add None to tests

File: yt_dlp/extractor/mxplayer.py [hunk 13]
```diff
@@ -1,178 +1,99 @@
...
+            if season:
+                match = re.match(r'Season (?P<season_number>\d+)', season)
+                if match:
+                    season_number = int(match.group('season_number'))
+
+        for key in 'dash', 'hls':
+            playlist_url = STREAM_URL.format(data_json['stream'][key]['high'])
+
+            if key == 'dash':
+                frmt, subs = self._extract_mpd_formats_and_subtitles(playlist_url, display_id, fatal=False)
```
[the remark this answers is not in the document]
Comment (itachi-19) [reply yt_dlp/extractor/mxplayer.py]: Done.

File: yt_dlp/extractor/mxplayer.py (hunk 8 above)
Comment (itachi-19) [reply yt_dlp/extractor/mxplayer.py:177]: Done.

File: yt_dlp/extractor/mxplayer.py (hunk 9 above)
Comment (itachi-19) [reply yt_dlp/extractor/mxplayer.py:167]: Done.

File: yt_dlp/extractor/mxplayer.py (hunk 10 above)
Comment (itachi-19) [reply yt_dlp/extractor/mxplayer.py:180]: Ah..Looks clean. Done.

File: yt_dlp/extractor/mxplayer.py (hunk 11 above)
Comment (itachi-19) [reply yt_dlp/extractor/mxplayer.py:16]: Fetching the description from description tag in json. It typically has some form of description of the video. Currently I couldn't find md5 anywhere in the json being fetched.

File: yt_dlp/extractor/mxplayer.py (hunk 12 above)
Comment (itachi-19) [reply yt_dlp/extractor/mxplayer.py:37]: Got it. Done.

File: yt_dlp/extractor/mxplayer.py [hunk 14]
```diff
@@ -1,6 +1,10 @@
 from .common import InfoExtractor
 from ..compat import compat_str
-from ..utils import try_get
+from ..utils import (
+    try_get,
+    traverse_obj,
+    int_or_none,
+)
```
Comment (HobbyistDev) [inline yt_dlp/extractor/mxplayer.py:7]: sort

File: yt_dlp/extractor/mxplayer.py [hunk 15]
```diff
@@ -106,73 +106,93 @@ class MxplayerIE(InfoExtractor):
...
+                frmts, subs = self._extract_m3u8_formats_and_subtitles(playlist_url, display_id, fatal=False)
+
+            formats.extend(frmts)
+            self._merge_subtitles(subs, target=subtitles)
+
         self._sort_formats(formats)
+
         return {
             'id': video_id,
+            'title': data_json['title'],
```
Comment (HobbyistDev) [inline yt_dlp/extractor/mxplayer.py:186]: Title shouldn't fatal
```suggestion
            'title': data_json.get('title'),
```

File: yt_dlp/extractor/mxplayer.py [hunk 16]
```diff
@@ -106,73 +106,93 @@ class MxplayerIE(InfoExtractor):
...
-            elif dct.get('type') == 'tvshow_show':
-                series = dct.get('name')
-        thumbnails = []
-        for thumb in data_json.get('poster', []):
-            thumbnails.append({
-                'url': thumb.get('url'),
-                'width': thumb.get('width'),
-                'height': thumb.get('height'),
-            })
+        video_type, display_id, video_id = self._match_valid_url(url).groups()
```
Comment (HobbyistDev) [inline yt_dlp/extractor/mxplayer.py:153]: use `group` instead `groups`
```suggestion
        video_type, display_id, video_id = self._match_valid_url(url).group('type', 'display_id', 'id')
```

File: yt_dlp/extractor/mxplayer.py (hunk 11 above)
Comment (elyse0) [reply yt_dlp/extractor/mxplayer.py:19]: The md5 checksum is computed by the test runner, taking the field as input (in this case "description"). As pukkandan said, in this case is okay because it's short but if it were a description with >150 characters we would prefer md5

https://www.md5hashgenerator.com/
image

File: yt_dlp/extractor/mxplayer.py (hunk 15 above)
Comment (itachi-19) [reply yt_dlp/extractor/mxplayer.py:186]: 'title' field is mandatory AFAIK. So I believe it doesn't make sense to make it non-fatal.

File: yt_dlp/extractor/mxplayer.py (hunk 14 above)
Comment (itachi-19) [reply yt_dlp/extractor/mxplayer.py:7]: Done.

File: yt_dlp/extractor/mxplayer.py (hunk 15 above)
Comment (pukkandan) [reply yt_dlp/extractor/mxplayer.py:186]: https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#mandatory-and-optional-metafields
> While all extractors must return a `title`, they must also allow it's extraction to be non-fatal

Comment (itachi-19) [reply yt_dlp/extractor/mxplayer.py:186]: Oh. My bad for not reading it fully. Done.

File: yt_dlp/extractor/mxplayer.py (hunk 11 above)
Comment (itachi-19) [reply yt_dlp/extractor/mxplayer.py:19]: Ok. So should I change all descriptions to md5 hash? I believe there are some descriptions longer than 150.

Comment (itachi-19) [reply yt_dlp/extractor/mxplayer.py:19]: Ok. This is done. Have replaced all descriptions for the extractor with md5 hashes.

File: yt_dlp/extractor/mxplayer.py (hunk 8 above)
Comment (itachi-19) [reply yt_dlp/extractor/mxplayer.py]: This is done. Had to use an if-else for avoiding return of 0 as episode number and instead return None.

File: yt_dlp/extractor/mxplayer.py [hunk 17]
```diff
@@ -107,72 +107,88 @@ class MxplayerIE(InfoExtractor):
...
-                frmt, subs = self._extract_mpd_formats_and_subtitles(dct.get('playUrl'), display_id, fatal=False)
-                formats.extend(frmt)
-                subtitles = self._merge_subtitles(subtitles, subs)
-            elif dct.get('extension') == 'm3u8':
-                frmt, subs = self._extract_m3u8_formats_and_subtitles(dct.get('playUrl'), display_id, fatal=False)
-                formats.extend(frmt)
-                subtitles = self._merge_subtitles(subtitles, subs)
+
+        season = traverse_obj(data_json, ('container', 'title'))
+        episode_number = data_json.get('sequence')
```
Comment (pukkandan) [inline yt_dlp/extractor/mxplayer.py:165]: inline in the return

**Review by pukkandan — APPROVED** [review]

File: yt_dlp/extractor/mxplayer.py (hunk 17 above)
Comment (pukkandan) [reply yt_dlp/extractor/mxplayer.py:165]:
```suggestion
        season = traverse_obj(data_json, ('container', 'title'))
        episode_number = data_json.get('sequence')
```
